Image Component (#229)
primer/brand: apps/docs/content/components/Image.mdx

---
title: Image
status: Experimental
source: https://github.com/primer/brand/blob/main/packages/react/src/Image/Image.tsx
storybook: '/brand/storybook/?path=/docs/components-image--playground'
description: Use Image to create both normal and responsive image elements.
---

import {Label} from '@primer/react'

```js
import {Image} from '@primer/react-brand'
```

## Examples

### Default

```jsx live
<Image src="https://source.unsplash.com/random" alt="Random Image" />
```

### Picture

The `as` prop can be used to set the container of the image to be use `picture`.

```jsx live

> <Image

Reviewer: If I wanted to generate this markup:

```html
<picture>
  <source srcset="logo-768.png, logo-768-1.5x.png 1.5x" />
  <img src="logo-320.png" alt="logo" />
</picture>
```

How would I do that with as="picture"? I can't see a way to set `sources`.

Author: Currently it sets on the image tag but I could include the children prop. Do you think that would be beneficial? I wasn't sure if that was really more performant.

Reviewer: We'd deffo need a way to produce the `<source>` elements when setting `as="picture"`. I don't think we need to do it via JSX children, but it should render markup like the example above. Could you show me me how you saw it working currently without doing that?

Author: Basically you would apply the property to the image. What I can do it parse the property from the img component and then take each value to create a source component. Would that work for you?

| `src`         | `string`                                |             | `true`   | Specifies the path to the image                                                                                                                                           |
| `alt`         | `string`                                |             | `true`   | Specifies a text value explaining the nature of the image for users of Assistive Technolog                                                                                                                         y
| `as`          | `img`, `picture`                        | `img`       | `false`  | Specification to create a picture component                                                                                                                               |
| `aspectRatio` | `'1:1'`, `'16:9'`, `'16:10'`, `'4:3'`, `'custom'` | `undefined` | `false`  | Sets the image aspect ratio. A custom ratio can be provided in the design tokens.                                                                                         |

> | `height`      | `number`                                |             | `false`  | The height of the image element or its container if it has an aspect ratio                                                                                                |

Reviewer: FYI The table seems to be breaking around here somewhere...

![Screenshot 2023-04-17 at 09 31 50](https://user-images.githubusercontent.com/13340707/232432077-71341f42-c844-414a-9934-89e75fc423b2.png)

Author: this should be resolved in this build

Reviewer: I'm still seeing it unfortunately and can't see any upcoming new deployments.. when you say 'this build', has it been pushed up?

Author: "This build" means the current build pushed to GitHub for me. It seems by local prettier is adding a space so it keeps adding and removing. It should be fixed now in the newest build.

---
title: Image
status: Experimental

> # figma: https://www.figma.com/file/BJ95AjraesmRCWsKA013GS/Primer-Brand?node-id=539%3A14285&t=RJEWZhVUtrRQ0pKh-0

Reviewer: This can be removed altogether if it's not applicable

```suggestion
```

source: https://github.com/primer/brand/blob/main/packages/react/src/Image/Image.tsx
storybook: '/brand/storybook/?path=/docs/components-image--playground'
description: Use Image to create both normal and responsive image elements.
---

import {Label} from '@primer/react'

```js
import {Image} from '@primer/react-brand'
```

## Examples

### Default

```jsx live
<Image src="https://source.unsplash.com/random" alt="Random Image" />
```

### Picture

> The `as` prop can be used to set the container of the image to be a picture component.

Reviewer: ```suggestion
The `as` prop can be used to set the container of the image to be use `picture`.
```

```jsx live
<Image
  src="https://source.unsplash.com/random"
  alt="Random Image"

>   as="picture"

Reviewer: could we hoist as="picture" to the top as it's most relevant to this example?

/>
```

### Aspect ratio

The `aspectRatio` prop can be used to set the aspect ratio of the image. This is useful when the image is not the same aspect ratio as the container.

```jsx live
<Image
  src="https://source.unsplash.com/random"
  alt="Random Image"
  aspectRatio="16:9"
/>
```

### Height

The `height` prop can be used to set the height of the image. This can be used along side the `aspectRatio` prop to create a responsive image the same size as other images.

```jsx live
<Image
  src="https://source.unsplash.com/random"
  alt="Random Image"

>   height={200}

Reviewer: Not sure if this example is working 🤔 ... changing the height value isn't updating the live code example, it's always rendering full-height.

  aspectRatio="16:9"
/>
```

### Width

The `width` prop can be used to set the width of the image. This can be used along side the `aspectRatio` prop to create a responsive image the same size as other images.

```jsx live
<Image

>   src="https://source.unsplash.com/random"

Author: I think we discussed this before but I would like to use a different color for this image because I can't see this on my monitor. So I will pick the next darker grey available.

  alt="Random Image"
  width={200}

>   aspectRatio="16:9"

Reviewer: The aspect ratio here doesn't seem to be 16:9, is it working correctly? It looks more square than rectangular.

![Screenshot 2023-04-14 at 11 13 10](https://user-images.githubusercontent.com/13340707/232017218-35ee40f0-e728-4fea-bedb-ebc5a0ad555f.png)

/>
```

## Component props

### Image <Label>Required</Label>

> The first row in a table is automatically styled as a table header row.

Reviewer: ☝️ Is this intentional? Looks like it's from [ComparisonTable](https://primer-9770a149af-26139705.drafts.github.io/components/ComparisonTable#comparisontablerow-required). If so, can we remove this please?

```suggestion
```

| name          | type                                    | default     | required | description                                                                                                                                                               |
| ------------- | --------------------------------------- | ----------- | -------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `src`         | `string`                                |             | `true`   | Specifies the path to the image                                                                                                                                           |

> | `alt`         | `string`                                |             | `true`   | Specifies an alternate text for the image                                                                                                                                 |

Reviewer: ```suggestion
| `alt`         | `string`                                |             | `true`   | Specifies a text value explaining the nature of the image for users of Assistive Technolog                                                                                                                         y
```

| `as`          | `img`, `picture`                        | `img`       | `false`  | Specification to create a picture component                                                                                                                               |

> | `aspectRatio` | `1:1`, `16:9`, `16:10`, `4:3`, `custom` | `undefined` | `false`  | Sets the image aspect ratio. A custom ratio can be provided in the design tokens.                                                                                         |

Reviewer: Is `custom` still a valid value, or were we dropping it. If it's valid, can we add an example of how users should use it please?

Reviewer: Could we add quotes around these to convey these are strings?

```suggestion
| `aspectRatio` | `'1:1'`, `'16:9'`, `'16:10'`, `'4:3'`, `'custom'` | `undefined` | `false`  | Sets the image aspect ratio. A custom ratio can be provided in the design tokens.                                                                                         |

````

---
title: Image
status: Experimental
source: https://github.com/primer/brand/blob/main/packages/react/src/Image/Image.tsx
storybook: '/brand/storybook/?path=/docs/components-image--playground'
description: Use Image to create both normal and responsive image elements.
---

import {Label} from '@primer/react'

```js
import {Image} from '@primer/react-brand'
```

## Examples

### Default

This component uses the `img` element by default.

```jsx live
<Image
  src="https://via.placeholder.com/600x400/d3d9df/d3d9df.png"
  alt="placeholder, blank area with an off-white background color"
/>
```

> 

Reviewer: Could we add an additional example using `srcSet` on an `img` please? That would be a super-helpful for copy-pasting.

Author: This has been added

### Picture

> The `as` prop can be used to set the container of the image to be use `picture`.

Reviewer: Typo

```suggestion
The `as` prop can be used to set the container of the image to use `picture`.
```

Author: This has been applied

```jsx live
<Image

>   as="picture"

Reviewer: This example is setting an empty `source`. Could we add a test to guard this please? Should only appear if a `source` is defined.

![Screenshot 2023-04-18 at 14 32 23](https://user-images.githubusercontent.com/13340707/232793425-a23d64d2-fcb4-4ed6-811d-e7ccfb7c72e4.png)

I assume most users of `as="picture"` would want to set some sources manually. Could we update this example to show how you would do this? I'm not quite clear on that myself right now either.

Author: This would require access to children if we want to do that simply. I can include that but it would overcomplicate the component a bit possibly.

Author: I think we should try to avoid doing this because it is going to complicate things quite a bit possibly.

Author: This has been updated to work when defined but not a custom setting.

Author: This has be discussed and resolved here: https://github.com/primer/brand/pull/229#discussion_r1171768591
